### ISSUE TITLE
Spatial index builder not indexing all records

### ISSUE THREAD

**Issue by johnvanbreda** [BiologicalRecordsCentre/iRecord#220]
There are gaps in the indexation of records by the spatial index builder - some records are not being lodged in index_locations_samples or linked via the location_id_* fields in the cache tables even though they should be.

**Comment by johnvanbreda** [comment]
By running the scheduled tasks manually, I have discovered the following error is being emitted sometimes, which presumably is the result of an invalid location boundary preventing an entire "batch" from indexing:
```
There was an SQL error: ERROR: GEOSTouches: TopologyException: side location conflict at -363168.08227949601 7047577.8439409798 - 
select distinct 
      l.id, s.id, coalesce(linked.id, 0) = l.id or st_contains(l.boundary_geom, s.geom), l.location_type_id
    from locations l
    join cache_termlists_terms t on t.id=l.location_type_id
    join samples s on s.deleted=false
      and (st_intersects(l.boundary_geom, s.geom) and not st_touches(l.boundary_geom, s.geom))
    join smplist list on list.id=s.id
    left join index_locations_samples ils on ils.location_id=l.id and ils.sample_id=s.id
    join cache_samples_nonfunctional snf on snf.id=s.id
    left join locations linked on linked.id=snf.attr_linked_location_id and linked.deleted=false
    where ils.id is null
    and l.deleted=false
    and (
      -- if a linked_location_id specified, then limit the indexing to the linked location for this location type.
      snf.attr_linked_location_id is null 
      or linked.id = l.id 
      or linked.location_type_id &lt;&gt; l.location_type_id 
      )
    and t.preferred_term in ('Vice County','LRC Boundary','Country','Cairngorms NP NNR','Miscellaneous indexed boundaries','Butterfly Conservation Branch','NT Bioblitz 2015 site')
    and (t.preferred_term&lt;&gt;'NT Bioblitz 2015 site' or s.survey_id in (257))
```

**Comment by kitenetter** [comment]
@johnvanbreda is this likely to be a result of continuing problems with the LERC boundaries? I will see if ALERC have been able to progress the updated GIS layer.

**Comment by johnvanbreda** [comment]
I'm looking into it now - I haven't located the exact problem but so far I don't think it's an LRC boundary that causes the issue.

**Comment by johnvanbreda** [comment]
2 problems identified:
1) Where there are broken polygons or multipolygons in the locations table, sometimes the queries fail leaving batches of records unindexed. This occurs even if the location is not on one of the indexed layers. I've deployed a fix so that as long as the indexed layers do not contain any problems then it will work OK.
2) If an occurrence is added to an existing sample that has already been indexed then it does not get added to the index. I've also added a fix for this.

I've also run a script to fix all the missing indexing in the data, but will double check in a day or two that the fixes are working before closing this.
